src/ray/gcs/gcs_server/gcs_actor_distribution.cc

```diff
@@ -0,0 +1,669 @@
+  ostr << indent_1 << "language = " << rpc::Language_Name(language_) << ",\n";
+  ostr << indent_1 << "is_shared = " << is_shared_ << ",\n";
+  ostr << indent_1 << "slot_capacity = " << slot_capacity_ << ",\n";
+  ostr << indent_1 << "available_slot_count = " << GetAvailableSlotCount() << ",\n";
+  ostr << indent_1 << "acquired_resources = " << acquired_resources_.ToString() << "\n";
+  ostr << indent_0 << "},";
+  return ostr.str();
+}
+
+//////////////////////////////////////////////////////////////////////
+GcsJobSchedulingContext::GcsJobSchedulingContext(const GcsJobConfig &job_config)
+    : job_config_(job_config) {
+  std::unordered_map<std::string, double> resource_mapping;
+  resource_mapping.emplace(kMemory_ResourceLabel, job_config_.total_memory_units_);
+  scheduling_resources_ = SchedulingResources(ResourceSet(resource_mapping));
+
+  for (uint32_t i = 0; i < job_config.num_initial_java_worker_processes_; ++i) {
+    RAY_CHECK(AddDummySharedActorWorkerAssignment());
+  }
+}
+
+bool GcsJobSchedulingContext::AddDummySharedActorWorkerAssignment() {
+  std::unordered_map<std::string, FixedPoint> resource_mapping;
+  resource_mapping.emplace(kMemory_ResourceLabel,
+                           job_config_.java_worker_process_default_memory_units_);
+  ResourceSet constraint_resources(resource_mapping);
+  auto actor_worker_assignment = GcsActorWorkerAssignment::Create(
+      NodeID::Nil(), job_config_.job_id_, Language::JAVA, constraint_resources,
+      /*is_shared=*/true, job_config_.num_java_workers_per_process_);
+  return AddActorWorkerAssignment(std::move(actor_worker_assignment));
+}
+
+bool GcsJobSchedulingContext::AddActorWorkerAssignment(
+    std::shared_ptr<GcsActorWorkerAssignment> actor_worker_assignment) {
+  RAY_CHECK(actor_worker_assignment != nullptr);
+
+  if (actor_worker_assignment->IsShared()) {
+    shared_actor_worker_assignments_.emplace(
+        actor_worker_assignment->GetActorWorkerAssignmentID(), actor_worker_assignment);
+  } else {
+    sole_actor_worker_assignments_.emplace(
+        actor_worker_assignment->GetActorWorkerAssignmentID(), actor_worker_assignment);
+  }
+
+  // Update the distribution of actor worker assignment on each node.
+  UpdateNodeToActorWorkerAssignment(actor_worker_assignment);
+
+  // TODO(Chong-Li):
+  // Acquire resources from job scheduling resources.
+  // The resources should be return back to the `scheduling_resources_` when the worker
+  // process is removed.
+
+  return true;
+}
+
+const GcsJobConfig &GcsJobSchedulingContext::GetJobConfig() const { return job_config_; }
+
+const ActorWorkerAssignmentMap &GcsJobSchedulingContext::GetSharedActorWorkerAssignments()
+    const {
+  return shared_actor_worker_assignments_;
+}
+
+const NodeToActorWorkerAssignmentsMap &
+GcsJobSchedulingContext::GetNodeToActorWorkerAssignments() const {
+  return node_to_actor_worker_assignments_;
+}
+
+bool GcsJobSchedulingContext::UpdateNodeToActorWorkerAssignment(
```

Comment:
some note on unit test:

we should have unit test to cover these branches,
- integration tests or e2e tests could rarely cover all branches.
- it also help the reviewer/reader to understand the semantics of each classes
- it's a forcing function to structure the code in a unit-testable way, usually results in cleaner interfaces.
- we should aim to cover all major interfaces and behavior of your classes, it really force you to hide details and leak as less as possible.